webExtensionApis/study/schema.yaml

```diff
@@ -0,0 +1,359 @@
+# possible types are here
+# - https://searchfox.org/mozilla-central/source/toolkit/components/extensions/Schemas.jsm#2418-2426
+#
+# - properties
+# - functions
+# - events
+# - types
+#
+# - type property:
+#   boolean, integer, number, string, array, object, or any.
+#
+
+# TODO
+# - add json schema types
+# - do we want to expose any of the sampling / hashing stuff?
+# - mention / decide how to handle all
+---
+- namespace: study
+  description: Interface for Shield and Pioneer studies.
+  apiVersion: 5  # for dev use
+  types:   # TODO @glind to actually get this back out of the schemas.json
+  - id: studySetup
+    type: object
+    properties:
+      activeExperimentName:
+        type: string
+      studyType:
+        type: "string"
+        enum: ["shield", "pioneer"]
+      expire:
+        type: 'object'
+        properties:
+          days:
+            type: 'integer'
+      endings:
+        type: "array"
+      weightedVariations:
+        type: "array"
+      logLevel:
+        type: "integer"
+        minimum: 0
+      telemetry:
+        type: object
+        properties:
+          send:
+            type: boolean
+          removeTestingFlag:
+            type: boolean
+
+    required:
+    - activeExperimentName
+    - studyType
+    - endings
+    - weightedVariations
+    - logLevel
+    - telemetry
+    additionalProperties: {}
+    testcase: {
+      activeExperimentName: aStudy,
+      studyType: 'shield',
+      expire: { "days": 10},
+      endings: [],
+      logLevel: 30,
+      weightedVariations: [{name: feature-active, weight: 1.5}],
+      telemetry: {  send: false, removeTestingFlag: false}
+    }
+
+  - id: weightedVariations
+    type: array
+    items:
+      type: 'object'
+      properties:
+        name:
+          type: "string"
+        weight:
+          type: "number"
+          minimum: 0
+      required: ["name", "weight"]
+    testcase: [{name: feature-active, weight: 1.5}]
+
+  - id: anEnding
+    type: object
+    additionalProperties: true
+    testcase: {baseUrls: ['some.url'], endingName: 'anEnding', endingClass: 'ended-positive'}
+
+  - id: filterTelemetryQuery
+    type: object
+    properties:
+      type: {type: ['array'], items: {type:'string'}}
+      n: {type: integer}
+      mininumTimestamp: {type: number}
+      headersOnly: {type: boolean}
+    testcase: {type: ['shield-study-addon', 'shield-study'], n: 100, minimumTimestamp: 1523968204184, headersOnly: false}
+
+  functions:
+  - name: setup
+    type: function
+    async: true
+    description: |
+      Attempt an setup/enrollment, with these effects:
+
+      - sets 'studyType' as Shield or Pioneer
+        - affects telemetry
+        - watches for dataPermission changes that should *always*
+          stop that kind of study
+
+      - Use or choose variation
+        - `testing.variation` if present
+        - OR deterministicVariation
+          for the studyType using `weightedVariations`
+
+      - During firstRun[1] only:
```

Comment:
(All arguments are validated as part of being a wee.  Are you suggesting we make that clearer / offer a tool?  Because that might be worth a bug, since these are complex objects.)